mapit_gb/management/commands/mapit_UK_add_2020_structural_changes.py

```diff
@@ -0,0 +1,72 @@
+# coding=utf8
+
+from django.core.management.base import BaseCommand
+from mapit.models import Area, Generation, Type, Country, NameType, CodeType
+
+
+class Command(BaseCommand):
+    help = "Sort out boundaries for 2020 council changes"
+
+    def add_arguments(self, parser):
+        parser.add_argument('--commit', action='store_true')
+
+    def handle(self, *args, **options):
+        self.commit = options['commit']
+        self.g = Generation.objects.current()
+        self.gn = Generation.objects.new()
+        self.name_type = NameType.objects.get(code='O')
+        self.code_type = CodeType.objects.get(code='gss')
+        self.country = Country.objects.get(code='E')
+
+        self.raise_generation_on_everything_else()
+
+        self.areas = Area.objects.filter(generation_low_id__lte=self.g, generation_high_id__gte=self.g)
+
+        self.get_existing()
+        self.create_buckinghamshire()
+
+    def _exclude_councils_and_wards(self, qs, names, type_p, type_c):
+        areas = Area.objects.filter(type__code=type_p, name__in=names, generation_high=self.g)
+        wards = Area.objects.filter(type__code=type_c, parent_area__in=areas, generation_high=self.g)
+        qs = qs.exclude(id__in=areas).exclude(id__in=wards)
+        return qs
+
+    def raise_generation_on_everything_else(self):
+        qs = Area.objects.filter(generation_high=self.g)
+        print('%d areas in database' % qs.count())
+        qs = self._exclude_councils_and_wards(qs, ('Buckinghamshire County Council',), 'CTY', 'CED')
+        qs = self._exclude_councils_and_wards(
+            qs, ('Aylesbury Vale District Council', 'South Buckinghamshire District Council',
+                 'Wycombe District Council', 'Chiltern District Council'), 'DIS', 'DIW')
+
+        if self.commit:
+            print('Raising gen high on %d areas' % qs.count())
+            updated = qs.update(generation_high=self.gn)
+            print('Raised gen high on %d areas' % updated)
+        else:
+            print('Would raise gen high on %d areas' % qs.count())
+
+    def get_existing(self):
+        self.buckinghamshire = Area.objects.get(type__code='CTY', name='Buckinghamshire County Council')
+
+    def _create(self, name, typ, area, gss=None):
+        assert(area.polygons.count() == 1)
+        geom = area.polygons[0].polygon
+        m = Area(
+            name=name, type=Type.objects.get(code=typ), country=self.country,
+            generation_low=self.gn, generation_high=self.gn,
+        )
+        if self.commit:
+            m.save()
+            m.names.update_or_create(type=self.name_type, defaults={'name': name})
+            if gss:
+                m.codes.update_or_create(type=self.code_type, defaults={'code': gss})
+            m.polygons.create(polygon=geom)
```

Comment:
Running locally with `--commit`, this line crashes with
```
psycopg2.errors.CheckViolation: new row for relation "mapit_geometry" violates check constraint "enforce_srid_polygon"
DETAIL:  Failing row contains (996813, 163793, 0103000020E6100000010000009E3C0000EDF23EFC23E7ECBF15565944DFC549...).
```

Reply:
I don't understand this one, the polygon has just come out of the db so how can it be a different SRID? :-/ What is your DB srid set to?

Reply:
Running on server (without commit), the geom's srid appears to be 27700 when fetched so should be okay on saving. I guess if your db mismatched your general.yml this could happen? But if it's consistent, will have to keep looking.

Reply:
Sigh, I had it `AREA_SRID: 4326` in general.yml. Sorry for the wild goose chase. Works fine now!